### PR TITLE
[DX] Make JsonThrowOnErrorRector work with local scope to detect error_json methods

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

File: rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php [hunk 1]
```diff
@@ -49,24 +52,48 @@ public function getRuleDefinition(): RuleDefinition
...
+        );
+
+        if ($hasJsonErrorFuncCall) {
             return null;
         }
 
-        if ($this->isName($node, 'json_encode')) {
-            return $this->processJsonEncode($node);
-        }
+        $this->hasChanged = false;
```
Comment (samsonasik) [inline rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: Rector rule is shared service, so `$this->hasChanged` should not be used, as it marking all next skipped this rule as changed, local `$hasChanged` should be used instead .

Comment (TomasVotruba) [reply rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: This should be ok, as `hasChanged` is local private property only in current Rector rule.

Comment (samsonasik) [reply rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: on same Rector re-called, unfortunatelly, `$this->hasChanged` still alive on next same rector rule call

Comment (TomasVotruba) [reply rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: Not sure I see it. Could you add a failing test fixture?

Comment (samsonasik) [reply rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: this is happen when:

- same rector rule called 
- jump to different file

Comment (samsonasik) [reply rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: that's need e2e test as the change is about showing:

```
Applied rules:
 * JsonThrowOnErrorRector
```

Comment (samsonasik) [reply rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: Oh, I see, `$this->hasChanged = false` always init inside `refactor()`, that's why it is working as expected.

The not working part is when we define `private bool $hasChanged = false` without re-init in `refactor()` method.

so, this change is fine then ;)

Comment (TomasVotruba) [reply rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: Good to double-check though :), thank you :+1:

Comment (samsonasik) [reply rules/Php73/Rector/FuncCall/JsonThrowOnErrorRector.php:73]: I created PR https://github.com/rectorphp/rector-src/pull/4454 to apply init `$this->hasChanged = false;` on `refactor()` method.

**Pull request merged**